Replace ssh with net=host hostexec pod and kubectl exec in e2e tests where possible

sttts opened the pull request:
- added RunHostCmd to test/e2e/util.go to run a command in a shell in a hostexec container
- ported kubeproxy e2e tests to hostexec
  - speed-up by parallel pod launch
  - speed-up pod delete with zero grace period
  - fix timeouts with blocking udp tests
- ported privilegedPod e2e test to hostexec
- ported NodePort e2e tests to hostexec
- make all of them work with mesos/docker cluster

Prerequisites: https://github.com/kubernetes/kubernetes/pull/17101

_NOTE_: When #17101 is merged, commit 9e9df46 can be removed to use the gcr.io image.

sttts commented:
/cc @ArtfulCoder should be ready soon. Comments are welcome.

sttts commented:
/cc @jdef for review of contrib/mesos changes

sttts commented:
@k8s-bot test this please

sttts commented:
Shippable fun again.

jdef commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -85,6 +85,7 @@ apiserver:
     --service-node-port-range=30000-32767
     --cloud-provider=mesos
     --cloud-config=/opt/mesos-cloud.conf
+    --allow-privileged
```
is this a leftover from testing, or required to allow remote exec?

sttts replied:
This is default in upstream now. There are even e2e tests for this (which are fixed in this PR).

sttts replied:
To answer your specific question: it's not needed for remote exec, only for the PrivilegePod e2e tests.

jdef commented on test/e2e/kubeproxy.go:
```diff
@@ -241,24 +242,18 @@ func (config *KubeProxyTestConfig) dialFromContainer(protocol, containerIP, targ
 	var cmd string
 	if protocol == "udp" {
-		cmd = fmt.Sprintf("echo 'hostName' | nc -w 1 -u %s %d", targetIP, targetPort)
+		cmd = fmt.Sprintf("echo 'hostName' | timeout -t 3 nc -w 1 -u %s %d", targetIP, targetPort)
```
is that really a long enough timeout? worried about introducing flakes

sttts replied:
It should be bigger (or equal) than that of nc. I guess we can decrease it to 1 sec (like the `-w 1`).

jdef replied:
i was actually worried about it being long enough, not that it might be too long. if `nc` is timing out after 1s then `timeout -t 3` is probably fine as-is

jdef commented on test/e2e/kubeproxy.go:
```diff
@@ -457,23 +469,35 @@ func (config *KubeProxyTestConfig) cleanup() {
 func (config *KubeProxyTestConfig) deleteNetProxyPod() {
 	pod := config.endpointPods[0]
-	config.getPodClient().Delete(pod.Name, nil)
+	config.getPodClient().Delete(pod.Name, api.NewDeleteOptions(0))
```
why is the force-delete needed here?

sttts replied:
Why not? It slows down tests without it.

jdef replied:
no specific reason as to "why not" -- just wondering why the change was made at all

jdef commented on test/e2e/service.go:
```diff
@@ -387,17 +387,11 @@ var _ = Describe("Services", func() {
-				Failf("expected node port (%d) to be in use", nodePort)
-			}
+		hostExec := LaunchHostExecPod(f.Client, f.Namespace.Name)
+		cmd := fmt.Sprintf(`ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN`, nodePort)
```
never heard of `ss`; learned something new today :)

jdef replied:
`grep` before `tail`?

sttts replied:
the tail only skips a line. Should be fine.

jdef commented on test/e2e/util.go:
```diff
@@ -1776,7 +1792,12 @@ func NodeSSHHosts(c *client.Client) ([]string, error) {
+
+	sshHosts := make([]string, 0, len(hosts))
+	for _, h := range hosts {
+		sshHosts = append(sshHosts, h+":22")
```
use net.JoinHostPort here?

sttts replied:
good point

jdef commented on test/images/hostexec/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+run apk --update add curl netcat-openbsd iproute2 && rm -rf /var/cache/apk/*
+
+CMD [ "sleep", "3600" ]
```
how did we come up with this time? if the intent is to block until killed can we a bit more explicit? e.g.
```sh
rm -f /fifo && mkfifo /fifo && exec cat </fifo
```

sttts replied:
Well, it's a test image. Hopefully a test does not take an hour. But I will gratefully use your perfect wait command :)

jdef commented on test/e2e/service.go:
```diff
@@ -742,17 +736,11 @@ var _ = Describe("Services", func() {
-				Failf("expected node port (%d) to not be in use", nodePort)
-			}
+		hostExec := LaunchHostExecPod(f.Client, f.Namespace.Name)
+		cmd := fmt.Sprintf(`! ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN`, nodePort)
```
don't we want to `grep` before `tail`?

sttts replied:
As above. It's not a `tail -f`. Moreover, I didn't want to touch the actual test.

jdef commented on test/images/hostexec/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+FROM alpine:3.2
+
+run apk --update add curl netcat-openbsd iproute2 && rm -rf /var/cache/apk/*
```
why is iproute2 needed?

sttts replied:
For our nice little tool `ss`. Will add a comment.

jdef commented on test/e2e/kubeproxy.go:
```diff
@@ -494,11 +518,6 @@ func (config *KubeProxyTestConfig) createPod(pod *api.Pod) *api.Pod {
 	if err != nil {
```
why is this an if statement, vs `Expect(err).NotTo(HaveOccurred())`? just wondering..

sttts replied:
Seems like this style is used in the whole file.

jdef commented on test/e2e/kubeproxy.go:
```diff
@@ -421,13 +430,16 @@ func (config *KubeProxyTestConfig) setup() {
+	nodeList, err := config.f.Client.Nodes().List(labels.Everything(), fields.Everything())
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Failed to get node list: %v", err))
+	config.externalAddrs = NodeAddresses(nodeList, api.NodeExternalIP)
+	if len(config.externalAddrs) < 2 {
```
assuming that this is what allows the test to run on k8s-mesos?

sttts replied:
Yeah, ssh only worked on gce, gke, aws.

sttts commented:
@k8s-bot test this please

sttts commented:
/cc @ArtfulCoder ready for (non-mesos) review

jdef commented:
@ArtfulCoder PTAL

ArtfulCoder commented on test/e2e/util.go:
```diff
@@ -1818,6 +1840,50 @@ func sshCore(cmd, host, provider string, verbose bool) (string, string, int, err
+			Containers: []api.Container{
+				{
+					Name:            "hostexec",
+					Image:           "sttts/hostexec:1.2",
```
We have to make sure that we change this image to gcr hosted one, before checking in...

Could we split this PR into two where the hostexec image can go in first ?

sttts replied:
I have prepared a separate PR: https://github.com/kubernetes/kubernetes/pull/17101

ArtfulCoder commented:
I love what you have done here.
Could you split this PR in two, so that we can get hostexec in first ?

sttts commented:
Rebased after #17101 got merged.

sttts committed:
e2e: replace ssh with net=host hostexec pod and kubectl exec

- document needed packages in hostexec image
- add RunHostCmdOrDie
- kube-proxy e2e: port from ssh to hostexec
  - use preset NodeName to schedule test pods to different nodes
  - parallel launch of pods
  - port from ssh to hostexec
  - add timeout because nc might block on udp
  - delete test container without grace period
- PrivilegedPod e2e: port from ssh to hostexec
- NodePort e2e: port from ssh to hostexec
- cluster/mesos/docker: Enable privileged pods

sttts commented:
@k8s-bot test this

sttts commented:
@ArtfulCoder rebased and squashed

sttts commented:
Failing e2e test looks like a flake: "should schedule a pod w/two RW PDs both mounted to one container".

sttts commented:
@k8s-bot test this

sttts commented:
Squashed e2e test ports. Rebased to master.

sttts commented:
Rebased to latest master.

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -288,15 +281,12 @@ func (config *KubeProxyTestConfig) createNetShellPodSpec(podName string) *api.Po
-						{
-							Name:          "host",
-							ContainerPort: endpointHttpPort,
-							HostPort:      endpointHostPort,
```
how do we make sure that the pod is created on every host ?

sttts replied:
Compare the nodeName and line 479.

ArtfulCoder commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -86,6 +86,7 @@ apiserver:
     --service-node-port-range=30000-32767
```
Please squash all commits

sttts commented:
@ArtfulCoder squashed.

sttts commented:
@k8s-bot test this

>

sttts commented:
e2e broken again? "Cluster failed to initialize within 300 seconds". Exactly same code as before, only squashed.

sttts commented:
@k8s-bot test this

sttts commented:
@ArtfulCoder ptal

ArtfulCoder commented:
Thank you for the changes!

sttts commented:
@k8s-bot test this
